File: python/rewire/rules.py
"""
Rewire rule evaluation: schedule and alert-path constraint checking.

All evaluations return evidence-based results.
"""

from __future__ import annotations

import json
import time
from dataclasses import dataclass
from typing import Optional, Tuple, List, Dict, Any

Timestamp = int
Evidence = Dict[str, Any]
ViolationTuple = Tuple[str, str, Evidence]  # (code, message, evidence)


def now_i() -> Timestamp:
    """Current Unix timestamp as integer."""
    return int(time.time())


@dataclass(frozen=True)
class ScheduleParams:
    """Constraints for schedule-type expectations."""
    max_runtime_s: int   # 0 disables check
    min_spacing_s: int   # 0 disables check
    allow_overlap: bool


@dataclass(frozen=True)
class AlertPathParams:
    """Constraints for alert-path expectations."""
    ack_window_s: int     # time allowed to acknowledge
    test_interval_s: int  # how often to send synthetic test


def parse_params(exp_type: str, params_json: str) -> ScheduleParams | AlertPathParams:
    """Parse type-specific parameters from JSON."""
    obj = json.loads(params_json)
    if exp_type == "schedule":
        return ScheduleParams(
            max_runtime_s=int(obj.get("max_runtime_s", 0)),
            min_spacing_s=int(obj.get("min_spacing_s", 0)),
            allow_overlap=bool(obj.get("allow_overlap", False)),
        )
    if exp_type == "alert_path":
        return AlertPathParams(
            ack_window_s=int(obj["ack_window_s"]),
            test_interval_s=int(obj["test_interval_s"]),
        )
    raise ValueError(f"unknown expectation type: {exp_type}")


def schedule_evaluate(
    exp_row: Any, obs_rows_desc: List[Any]
) -> Tuple[List[ViolationTuple], List[str]]:
    """
    Evaluate schedule constraints against observations.

    Args:
        exp_row: Expectation row from database
        obs_rows_desc: Observations sorted by observed_at DESC (newest first)

    Returns:
        Tuple of (violations, codes_to_close)
        - violations: list of (code, message, evidence) tuples
        - codes_to_close: list of violation codes that should be closed
    """
    params = parse_params("schedule", exp_row["params_json"])
    expected = int(exp_row["expected_interval_s"])
    tol = int(exp_row["tolerance_s"])
    t = now_i()

    violations: List[ViolationTuple] = []
    close_codes: List[str] = []

    # Find most recent start
    last_start = next(
        (r for r in obs_rows_desc if r["kind"] == "start"), None
    )

    # Check: missed execution
    if last_start:
        age = t - int(last_start["observed_at"])
        if age > expected + tol:
            violations.append((
                "missed",
                f"Expected a start within {expected}s (+{tol}s); last start was {age}s ago.",
                {
                    "last_start_at": int(last_start["observed_at"]),
                    "age_s": age,
                    "expected_s": expected,
                    "tolerance_s": tol,
                },
            ))
        else:
            close_codes.append("missed")

    # Check: overlap / longrun
    if last_start:
        start_t = int(last_start["observed_at"])
        # Find end after this start
        newer_end = next(
            (r for r in obs_rows_desc
             if r["kind"] == "end" and int(r["observed_at"]) >= start_t),
            None
        )

        if newer_end is None:
            # Job may still be running
            run_for = t - start_t
            if params.max_runtime_s and run_for > params.max_runtime_s:
                violations.append((
                    "longrun",
                    f"Run exceeded max_runtime_s={params.max_runtime_s}; running for {run_for}s.",
                    {
                        "start_at": start_t,
                        "running_for_s": run_for,
                        "max_runtime_s": params.max_runtime_s,
                    },
                ))
            else:
                close_codes.append("longrun")

            # Check overlap (start without end while another running)
            if not params.allow_overlap:
                starts_without_end = [
                    r for r in obs_rows_desc if r["kind"] == "start"
                ]
                if len(starts_without_end) > 1:
                    second = starts_without_end[1]
                    if int(second["observed_at"]) < start_t:
                        # There's an earlier start that also has no end
                        violations.append((
                            "overlap",
                            "Detected overlapping runs.",
                            {
                                "newest_start_at": start_t,
                                "other_start_at": int(second["observed_at"]),
                            },
                        ))
                    else:
                        close_codes.append("overlap")
                else:
                    close_codes.append("overlap")
        else:
            # Job completed
            close_codes.extend(["longrun", "overlap"])

            # Check spacing
            if params.min_spacing_s:
                prev_end = next(
                    (r for r in obs_rows_desc
                     if r["kind"] == "end" and int(r["observed_at"]) < start_t),
                    None
                )
                if prev_end:
                    gap = start_t - int(prev_end["observed_at"])
                    if gap < params.min_spacing_s:
                        violations.append((
                            "spacing",
                            f"Start occurred {gap}s after previous end; min_spacing_s={params.min_spacing_s}.",
                            {
                                "gap_s": gap,
                                "min_spacing_s": params.min_spacing_s,
                                "prev_end_at": int(prev_end["observed_at"]),
                                "start_at": start_t,
                            },
                        ))
                    else:
                        close_codes.append("spacing")

    return violations, close_codes


def alertpath_should_send_test(
    exp_row: Any, last_any_obs_time: Optional[Timestamp]
) -> bool:
    """Determine if it's time to send a synthetic alert test."""
    params = parse_params("alert_path", exp_row["params_json"])
    if last_any_obs_time is None:
        return True
    return (now_i() - int(last_any_obs_time)) >= params.test_interval_s

File: clojure/src/rewire/rules.clj
(ns rewire.rules
  "Rewire rule evaluation: schedule and alert-path constraint checking.

   All evaluations return evidence-based results."
  (:require [clojure.data.json :as json]
            [rewire.db :as db]))

;; ============================================================
;; Parameter parsing
;; ============================================================

(defn parse-schedule-params
  "Parse schedule-type parameters."
  [params-json]
  (let [obj (json/read-str params-json :key-fn keyword)]
    {:max-runtime-s (or (:max_runtime_s obj) 0)
     :min-spacing-s (or (:min_spacing_s obj) 0)
     :allow-overlap (boolean (:allow_overlap obj))}))

(defn parse-alertpath-params
  "Parse alert-path parameters."
  [params-json]
  (let [obj (json/read-str params-json :key-fn keyword)]
    {:ack-window-s (:ack_window_s obj)
     :test-interval-s (:test_interval_s obj)}))

(defn parse-params
  "Parse type-specific parameters from JSON."
  [exp-type params-json]
  (case exp-type
    "schedule" (parse-schedule-params params-json)
    "alert_path" (parse-alertpath-params params-json)
    (throw (ex-info "Unknown expectation type" {:type exp-type}))))

;; ============================================================
;; Schedule evaluation
;; ============================================================

(defn- find-obs
  "Find first observation matching predicate."
  [obs-rows pred]
  (first (filter pred obs-rows)))

(defn schedule-evaluate
  "Evaluate schedule constraints against observations.

   Args:
     exp-row: Expectation row from database
     obs-rows-desc: Observations sorted by observed_at DESC (newest first)

   Returns:
     Map with :violations (list of [code message evidence]) and :close-codes"
  [exp-row obs-rows-desc]
  (let [params (parse-params "schedule" (:params_json exp-row))
        expected (:expected_interval_s exp-row)
        tol (:tolerance_s exp-row)
        t (db/now-i)
        violations (atom [])
        close-codes (atom [])]

    ;; Find most recent start
    (when-let [last-start (find-obs obs-rows-desc #(= (:kind %) "start"))]
      (let [start-t (:observed_at last-start)
            age (- t start-t)]

        ;; Check: missed execution
        (if (> age (+ expected tol))
          (swap! violations conj
                 ["missed"
                  (format "Expected a start within %ds (+%ds); last start was %ds ago."
                          expected tol age)
                  {:last_start_at start-t :age_s age
                   :expected_s expected :tolerance_s tol}])
          (swap! close-codes conj "missed"))

        ;; Find end after this start
        (let [newer-end (find-obs obs-rows-desc
                                  #(and (= (:kind %) "end")
                                        (>= (:observed_at %) start-t)))]
          (if (nil? newer-end)
            ;; Job may still be running
            (let [run-for (- t start-t)]
              (if (and (pos? (:max-runtime-s params))
                       (> run-for (:max-runtime-s params)))
                (swap! violations conj
                       ["longrun"
                        (format "Run exceeded max_runtime_s=%d; running for %ds."
                                (:max-runtime-s params) run-for)
                        {:start_at start-t :running_for_s run-for
                         :max_runtime_s (:max-runtime-s params)}])
                (swap! close-codes conj "longrun"))

              ;; Check overlap
              (when-not (:allow-overlap params)
                (let [starts (filter #(= (:kind %) "start") obs-rows-desc)]
                  (if (and (> (count starts) 1)
                           (< (:observed_at (second starts)) start-t))
                    (swap! violations conj
                           ["overlap"
                            "Detected overlapping runs."
                            {:newest_start_at start-t
                             :other_start_at (:observed_at (second starts))}])
                    (swap! close-codes conj "overlap")))))

            ;; Job completed
            (do
              (swap! close-codes conj "longrun" "overlap")

              ;; Check spacing
              (when (pos? (:min-spacing-s params))
                (when-let [prev-end (find-obs obs-rows-desc
                                              #(and (= (:kind %) "end")
                                                    (< (:observed_at %) start-t)))]
                  (let [gap (- start-t (:observed_at prev-end))]
                    (if (< gap (:min-spacing-s params))
                      (swap! violations conj
                             ["spacing"
                              (format "Start occurred %ds after previous end; min_spacing_s=%d."
                                      gap (:min-spacing-s params))
                              {:gap_s gap :min_spacing_s (:min-spacing-s params)
                               :prev_end_at (:observed_at prev-end) :start_at start-t}])
                      (swap! close-codes conj "spacing"))))))))))

    {:violations @violations
     :close-codes (vec (distinct @close-codes))}))

;; ============================================================
;; Alert path evaluation
;; ============================================================

(defn alertpath-should-send-test?
  "Determine if it's time to send a synthetic alert test."
  [exp-row last-any-obs-time]
  (let [params (parse-params "alert_path" (:params_json exp-row))]
    (or (nil? last-any-obs-time)
        (>= (- (db/now-i) last-any-obs-time) (:test-interval-s params)))))
